### PR TITLE
[FLINK-19092][sql-parser] Parse comment on computed column failed

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*Currently exception of parser is thrown when adding comment on computed column with DDL. Parser should support comment on computed column for DDL of `CREATE TABLE` as follows:*
```
<computed_column_definition>: column_name AS computed_column_expression [COMMENT column_comment]
```

## Brief change log

  - *Add `SqlTableComputedColumn` to support commented computed column in `Parser.tdd`.*
  - *`ComputedColumn ` of `parserImpls.ftl` which creates computed column returns the `SqlTableComputedColumn`.*
  - *`SqlCreateTable` modifies validation and unparse about the computed column by `SqlTableComputedColumn`.*
  - *`MergeTableLikeUtil` modifies the check for the computed column whether derivedColumn is instance of `SqlTableComputedColumn`.*

## Verifying this change

  - *`FlinkSqlParserImplTest` adds `testCreateTableWithCommentOnComputedColumn` to verify the test case that creates table with comment on the computed column.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wuchong** [comment]
cc @danny0405 , could you help to review this?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlTableComputedColumn.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	private SqlNode expr;
+	private SqlCharStringLiteral comment;
+
+	public SqlTableComputedColumn(SqlNode identifier, SqlNode expr, @Nullable SqlCharStringLiteral comment, SqlParserPos pos) {
+		super(pos);
+		this.identifier = requireNonNull(identifier, "Column identifier should not be null");
+		this.expr = requireNonNull(expr, "Column expression should not be null");
+		this.comment = comment;
+	}
+
```
Comment (danny0405) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlTableComputedColumn.java:55]: Can we merge this class into `SqlTableColumn` ? Just like what we do in schema `TableColumn`.

Comment (SteNicholas) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlTableComputedColumn.java]: @danny0405 Yes, I have already merged `SqlTableComputedColumn` into `SqlTableColumn`. Please review this pull request again.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlCreateTable.java [hunk 2]
```diff
@@ -236,12 +232,12 @@ public String getColumnSqlString() {
...
 		for (SqlNode column : columnList) {
 			writer.sep(",");
-			if (column instanceof SqlTableColumn) {
-				SqlTableColumn tableColumn = (SqlTableColumn) column;
-				tableColumn.getName().unparse(writer, 0, 0);
-			} else {
-				column.unparse(writer, 0, 0);
+			SqlTableColumn tableColumn = (SqlTableColumn) column;
+			if (tableColumn.isComputed()) {
+				tableColumn.getExpr().unparse(writer, 0, 0);
```
Comment (danny0405) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlCreateTable.java:237]: We can encapsulate all the unparse logic in `SqlTableColumn`. E.G. the code can be as simple as:
```java
column.unparse(writer, 0, 0)
```

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlTableColumn.java [hunk 3]
```diff
@@ -65,34 +67,55 @@ public SqlTableColumn(SqlIdentifier name,
...
 		}
 		if (this.comment != null) {
 			writer.print(" COMMENT ");
 			this.comment.unparse(writer, leftPrec, rightPrec);
 		}
 	}
 
+	public boolean isComputed() {
+		return type == null && expr != null;
+	}
```
Comment (danny0405) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlTableColumn.java:117]: - Can be `isGenerated` to be synced with `TableColumn`
- `return expr != null;` is enough

**Review by danny0405 — APPROVED** [review]
+1, looks good from my side.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlTableColumn.java [hunk 4]
```diff
@@ -109,6 +137,14 @@ public void setType(SqlDataTypeSpec type) {
 		this.type = type;
 	}
 
+	public SqlNode getExpr() {
```
Comment (wuchong) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlTableColumn.java:140]: Better to return `Optional<SqlNode>` to keep consistent with other nullable methods.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlTableColumn.java [hunk 5]
```diff
@@ -109,6 +137,14 @@ public void setType(SqlDataTypeSpec type) {
...
+		return expr;
+	}
+
+	public void setExpr(SqlNode expr) {
```
Comment (wuchong) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlTableColumn.java:144]: This is never used?

**Review by wuchong — COMMENTED** [review]
Looks good to me in general. I only left 2 minor comments.

**Review by wuchong — APPROVED** [review]
LGTM.

Waiting for the build to pass.

**Comment by wuchong** [comment]
Mering...

**Pull request merged**